doc/architecture.md

```diff
@@ -0,0 +1,206 @@
+
+> [!NOTE]
+> A new architecture is currently being planned for the code.  See [issue
+> #67](https://github.com/dandi/dandidav/issues/67) for more information.
+
+This document is best augmented by the doc comments in the `dandidav` source.
+You can view the doc comments in rendered form in a browser by running the
+following inside a clone of this repository:
+
+```console
+cargo doc --open --document-private-items --no-deps
+```
+
+General
+-------
+
+- `dandidav` is implemented as an [`axum`](https://github.com/tokio-rs/axum)
+  server in which almost all request-handling is done by a `Service`.  (We need
+  to use a `Service` rather than `axum`'s normal method routers because the
+  latter do not support WebDAV's `PROPFIND` verb.)  The service itself is
+  defined at [`src/main.rs`, lines 116-122][service-fn], and it simply passes
+  requests to the [`DandiDav::handle_request()`][handle-request] method for the
+  actual handling.
+
+    - The only requests not handled by the service are those for the CSS
+      stylesheet at `/.static/styles.css`, which is not WebDAV-enabled and thus
+      should not support `PROPFIND` or return the WebDAV-specific headers
+      present in all other responses.
+
+- If any error occurs during the processing of a request, it will almost always
+  "bubble up" to `DandiDav::handle_request()`, which will log the error and
+  convert it into either a 404 response or a 500 response, as appropriate.
+
+    - Panics (which are handled by `axum`) should only ever occur if there is
+      an actual bug in the code.
+
+- Most `GET` and `PROPFIND` requests to `dandidav` require fetching &
+  displaying information about both the resource identified by the request URL
+  and that resource's immediate children (if it has any); the exception is
+  `PROPFIND` requests with a `Depth` header of `0`, for which information about
+  child resources is not fetched.  In order to support both types of requests
+  efficiently, many methods & types come in two variants: one variant with the
+  suffix `_with_children()`/`WithChildren` for fetching or representing a
+  resource and its children, and one variant without the suffix for fetching or
+  representing just the resource itself.
+
+- File sizes are represented as `i64` instead of `u64` for compatibility with
+  the official Rust AWS S3 SDK.
+
+
+`DandiDav`
+----------
+
+- The [`DandiDav`][] type is responsible for handling WebDAV and plain HTTP
+  requests by:
+
+    - Parsing the request URL path into a [`DavPath`][] describing the type of
+      resource the URL refers to and the resource's parameters
+
+        - Splitting a path under a Dandiset version of the form
+          `{asset_path}/{zarr_resource_path}` is not done at this stage; that
+          is performed by the `DandiClient`.
+
+    - Making calls to a `DandiClient` or `ZarrManClient` instance as
+      appropriate to fetch information about the given resource and possibly
+      its child resources
```

Comment:
IIRC our endpoints for under zarr operations under `/dandisets/` as within the `.zarr` assets were notably slower than operation on zarrs under `/zarrs/`.  Is there any caching done/worth doing here to avoid redoing some splitting/decision making?

Reply:
There is no caching at the moment, but see #6.